File: worker/shared/secret_store.py
"""
shared/secret_store.py — Nơi DUY NHẤT worker đọc mật khẩu/token/khóa.

Bản Python của server/services/secret_store.js — cùng đọc danh mục
config/secrets_manifest.json và cùng thư mục secrets/:
  secrets/secrets.json            — mật khẩu EMR, token, API key...
  secrets/users.json              — tài khoản đăng nhập app
  secrets/nurse_emr_accounts.json — tài khoản EMR theo tên điều dưỡng

Thứ tự ưu tiên cho mỗi bí mật (cao → thấp):
  1. Biến môi trường (EMR_PASSWORD...) — server truyền tài khoản EMR riêng của
     người đang thao tác qua đây nên luôn phải thắng
  2. File chứa giá trị (EMR_PASSWORD_FILE=/đường/dẫn)
  3. secrets/secrets.json
  4. (Chỉ các khóa cũ) config/config.json — bị chặn khi EMR_REQUIRE_SECRET_ENV=1

Không đặt tên module là secrets.py để khỏi che module chuẩn `secrets`.
"""
from __future__ import annotations

import json
import os
import sys
from functools import lru_cache
from typing import Any, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
MANIFEST_PATH = os.path.join(ROOT_DIR, "config", "secrets_manifest.json")

SOURCE_ENV = "env"
SOURCE_ENV_FILE = "env_file"
SOURCE_SECRETS_FILE = "secrets_file"
SOURCE_LEGACY_CONFIG = "legacy_config"
SOURCE_NONE = ""


@lru_cache(maxsize=1)
def load_manifest() -> Dict[str, Any]:
    with open(MANIFEST_PATH, "r", encoding="utf-8") as f:
        raw = json.load(f)
    secrets = [s for s in (raw.get("secrets") or []) if isinstance(s, dict) and s.get("key")]
    files = [f for f in (raw.get("files") or []) if isinstance(f, dict) and f.get("name")]
    return {
        "secrets": secrets,
        "files": files,
        "by_key": {s["key"]: s for s in secrets},
        "file_by_name": {f["name"]: f for f in files},
    }


def _from_root(path: str) -> str:
    return path if os.path.isabs(path) else os.path.join(ROOT_DIR, path)


def secrets_dir() -> str:
    configured = str(os.environ.get("EMR_SECRETS_DIR", "") or "").strip()
    return _from_root(configured) if configured else os.path.join(ROOT_DIR, "secrets")


def secrets_file_path() -> str:
    return os.path.join(secrets_dir(), "secrets.json")


def _read_json_object(path: str) -> Dict[str, Any]:
    try:
        with open(path, "r", encoding="utf-8") as f:
            data = json.load(f)
    except FileNotFoundError:
        return {}
    except Exception as exc:
        print(f"[secret_store] {path} không đọc được JSON — bỏ qua: {exc}", file=sys.stderr)
        return {}
    return data if isinstance(data, dict) else {}


def read_secrets_file() -> Dict[str, Any]:
    return _read_json_object(secrets_file_path())


def _get_path(obj: Any, dotted: str) -> Any:
    cur = obj
    for part in str(dotted or "").split("."):
        if not isinstance(cur, dict):
            return None
        cur = cur.get(part)
    return cur


def _text(value: Any) -> str:
    return "" if value is None else str(value).strip()


def _read_env_file(env_name: str) -> str:
    path = _text(os.environ.get(f"{env_name}_FILE"))
    if not path:
        return ""
    try:
        with open(_from_root(path), "r", encoding="utf-8") as f:
            return f.read().strip()
    except Exception as exc:
        print(f"[secret_store] Không đọc được {env_name}_FILE: {exc}", file=sys.stderr)
        return ""


def resolve_secret(
    key: str,
    legacy_config: Optional[Dict[str, Any]] = None,
    secrets_data: Optional[Dict[str, Any]] = None,
) -> Tuple[str, str]:
    """Trả về (giá trị, nguồn). Không bao giờ log giá trị."""
    spec = load_manifest()["by_key"].get(key)
    if spec is None:
        raise KeyError(f"Bí mật chưa khai báo trong config/secrets_manifest.json: {key}")

    value = _text(os.environ.get(spec["env"]))
    if value:
        return value, SOURCE_ENV
    value = _read_env_file(spec["env"])
    if value:
        return value, SOURCE_ENV_FILE
    data = secrets_data if secrets_data is not None else read_secrets_file()
    value = _text(_get_path(data, spec.get("path", "")))
    if value:
        return value, SOURCE_SECRETS_FILE
    legacy_key = spec.get("legacy_config_key")
    if legacy_key and isinstance(legacy_config, dict):
        value = _text(legacy_config.get(legacy_key))
        if value:
            return value, SOURCE_LEGACY_CONFIG
    return "", SOURCE_NONE


def get_secret(key: str, default: str = "") -> str:
    value, _ = resolve_secret(key)
    return value or default


def apply_secrets_to_config(config: Dict[str, Any]) -> Dict[str, Any]:
    """Điền các khóa tài khoản EMR (username/password/hchanh_*/infusion_*) vào
    config đã merge, theo đúng thứ tự ưu tiên ở trên — code worker cũ vẫn đọc
    config["password"]... như trước, không phải sửa từng nơi."""
    out = dict(config or {})
    secrets_data = read_secrets_file()
    require_env = _text(os.environ.get("EMR_REQUIRE_SECRET_ENV")).lower() in {"1", "true", "yes", "on"}
    from_legacy = []
    for spec in load_manifest()["secrets"]:
        legacy_key = spec.get("legacy_config_key")
        if not legacy_key:
            continue
        value, source = resolve_secret(spec["key"], legacy_config=out, secrets_data=secrets_data)
        if source == SOURCE_LEGACY_CONFIG:
            from_legacy.append(legacy_key)
        elif value:
            out[legacy_key] = value
    if require_env and from_legacy:
        raise RuntimeError(
            "Mật khẩu dạng rõ trong config/config.json đã bị chặn bởi EMR_REQUIRE_SECRET_ENV. "
            f"Hãy chuyển các khóa sau sang secrets/secrets.json (npm run secrets:migrate) "
            f"hoặc biến môi trường: {', '.join(from_legacy)}"
        )
    return out


def resolve_secret_file(name: str) -> Tuple[str, str]:
    """Đường dẫn users.json / nurse_emr_accounts.json. Trả về (path, mode) với mode
    'env' | 'secrets' | 'legacy' — cùng quy tắc với secret_store.js."""
    spec = load_manifest()["file_by_name"].get(name)
    if spec is None:
        raise KeyError(f"File bí mật chưa khai báo trong config/secrets_manifest.json: {name}")
    configured = _text(os.environ.get(spec["env"]))
    if configured:
        return _from_root(configured), "env"
    in_secrets = os.path.join(secrets_dir(), name)
    if os.path.exists(in_secrets):
        return in_secrets, "secrets"
    legacy = _from_root(spec["legacy"]) if spec.get("legacy") else ""
    if legacy and os.path.exists(legacy):
        return legacy, "legacy"
    return in_secrets, "secrets"

File: server/services/secret_store.js
// server/services/secret_store.js — Nơi DUY NHẤT server đọc mật khẩu/token/khóa.
//
// Mọi bí mật của hệ thống gom về thư mục secrets/ (đã .gitignore, bị loại khỏi
// gói release):
//   secrets/secrets.json            — mật khẩu EMR, token app, salt, API key...
//   secrets/users.json              — tài khoản đăng nhập app
//   secrets/nurse_emr_accounts.json — tài khoản EMR theo tên điều dưỡng
//
// Danh mục khóa nằm ở config/secrets_manifest.json, dùng chung với
// worker/shared/secret_store.py — thêm bí mật mới thì khai ở đó, không đọc
// process.env trực tiếp rải rác trong code.
//
// Thứ tự ưu tiên cho mỗi bí mật (cao → thấp):
//   1. Biến môi trường, ví dụ EMR_APP_TOKEN
//   2. File chứa giá trị, ví dụ EMR_APP_TOKEN_FILE=/run/secrets/app_token
//   3. secrets/secrets.json
//   4. (Chỉ các khóa cũ) config/config.json — vẫn đọc được để không gãy máy
//      đang chạy, `npm run secrets:check` sẽ cảnh báo để chuyển sang secrets/.
//
// Module này KHÔNG require constants.js (constants.js require module này).

'use strict';

const fs = require('fs');
const path = require('path');

const ROOT_DIR = path.resolve(__dirname, '..', '..');
const MANIFEST_PATH = path.join(ROOT_DIR, 'config', 'secrets_manifest.json');
const LEGACY_CONFIG_PATH = path.join(ROOT_DIR, 'config', 'config.json');

const SOURCE = Object.freeze({
  ENV: 'env',
  ENV_FILE: 'env_file',
  SECRETS_FILE: 'secrets_file',
  LEGACY_CONFIG: 'legacy_config',
  NONE: '',
});

let manifestCache = null;
function loadManifest() {
  if (manifestCache) return manifestCache;
  const raw = JSON.parse(fs.readFileSync(MANIFEST_PATH, 'utf8'));
  const secrets = Array.isArray(raw.secrets) ? raw.secrets : [];
  const files = Array.isArray(raw.files) ? raw.files : [];
  manifestCache = Object.freeze({
    secrets: Object.freeze(secrets.map(s => Object.freeze({ ...s }))),
    files: Object.freeze(files.map(f => Object.freeze({ ...f }))),
    byKey: new Map(secrets.map(s => [s.key, s])),
    fileByName: new Map(files.map(f => [f.name, f])),
  });
  return manifestCache;
}

function resolveFromRoot(p) {
  return path.isAbsolute(p) ? p : path.join(ROOT_DIR, p);
}

// Đọc lại env mỗi lần gọi để test đổi EMR_SECRETS_DIR giữa chừng vẫn có tác dụng.
function secretsDir() {
  const configured = String(process.env.EMR_SECRETS_DIR || '').trim();
  return configured ? resolveFromRoot(configured) : path.join(ROOT_DIR, 'secrets');
}

function secretsFilePath() {
  return path.join(secretsDir(), 'secrets.json');
}

const warnedFiles = new Set();
function readJsonObject(filePath) {
  let text;
  try { text = fs.readFileSync(filePath, 'utf8'); } catch { return {}; }
  try {
    const parsed = JSON.parse(text);
    return parsed && typeof parsed === 'object' && !Array.isArray(parsed) ? parsed : {};
  } catch (err) {
    if (!warnedFiles.has(filePath)) {
      warnedFiles.add(filePath);
      console.error(`[secret_store] ${filePath} không phải JSON hợp lệ — bỏ qua: ${err.message}`);
    }
    return {};
  }
}

function getPath(obj, dotted) {
  let cur = obj;
  for (const part of String(dotted || '').split('.')) {
    if (!cur || typeof cur !== 'object') return undefined;
    cur = cur[part];
  }
  return cur;
}

function setPath(obj, dotted, value) {
  const parts = String(dotted || '').split('.');
  let cur = obj;
  for (const part of parts.slice(0, -1)) {
    if (!cur[part] || typeof cur[part] !== 'object' || Array.isArray(cur[part])) cur[part] = {};
    cur = cur[part];
  }
  cur[parts[parts.length - 1]] = value;
}

function text(value) {
  return value === undefined || value === null ? '' : String(value).trim();
}

function readEnvFile(envName) {
  const filePath = text(process.env[`${envName}_FILE`]);
  if (!filePath) return '';
  try {
    return fs.readFileSync(resolveFromRoot(filePath), 'utf8').trim();
  } catch (err) {
    console.error(`[secret_store] Không đọc được ${envName}_FILE: ${err.message}`);
    return '';
  }
}

/**
 * Tìm giá trị một bí mật. Trả về { value, source } — `source` cho biết giá trị
 * lấy từ đâu (để kiểm tra/diagnostics), KHÔNG bao giờ log `value`.
 * @param {string} key khóa trong config/secrets_manifest.json
 * @param {{secretsData?: object, legacyConfig?: object}} [opts] dữ liệu đã đọc sẵn (dùng khi duyệt nhiều khóa)
 */
function resolveSecret(key, opts = {}) {
  const spec = loadManifest().byKey.get(key);
  if (!spec) throw new Error(`[secret_store] Bí mật chưa khai báo trong config/secrets_manifest.json: ${key}`);

  const fromEnv = text(process.env[spec.env]);
  if (fromEnv) return { value: fromEnv, source: SOURCE.ENV };

  const fromEnvFile = readEnvFile(spec.env);
  if (fromEnvFile) return { value: fromEnvFile, source: SOURCE.ENV_FILE };

  const secretsData = opts.secretsData || readJsonObject(secretsFilePath());
  const fromSecrets = text(getPath(secretsData, spec.path));
  if (fromSecrets) return { value: fromSecrets, source: SOURCE.SECRETS_FILE };

  if (spec.legacy_config_key) {
    const legacy = opts.legacyConfig || readJsonObject(LEGACY_CONFIG_PATH);
    const fromLegacy = text(legacy[spec.legacy_config_key]);
    if (fromLegacy) return { value: fromLegacy, source: SOURCE.LEGACY_CONFIG };
  }
  return { value: '', source: SOURCE.NONE };
}

function getSecret(key) {
  return resolveSecret(key).value;
}

/**
 * Đường dẫn file danh sách bí mật (users.json, nurse_emr_accounts.json).
 *   1. Biến môi trường (EMR_USERS_FILE...)          → mode 'env'
 *   2. secrets/<name> đã tồn tại                    → mode 'secrets'
 *   3. Vị trí cũ config/<name> còn tồn tại           → mode 'legacy' (chưa chuyển)
 *   4. Chưa có ở đâu cả → tạo mới trong secrets/     → mode 'secrets'
 */
function resolveSecretFile(name) {
  const spec = loadManifest().fileByName.get(name);
  if (!spec) throw new Error(`[secret_store] File bí mật chưa khai báo trong config/secrets_manifest.json: ${name}`);
  const configured = text(process.env[spec.env]);
  if (configured) return { path: resolveFromRoot(configured), mode: 'env' };
  const inSecrets = path.join(secretsDir(), name);
  if (fs.existsSync(inSecrets)) return { path: inSecrets, mode: 'secrets' };
  const legacy = spec.legacy ? resolveFromRoot(spec.legacy) : '';
  if (legacy && fs.existsSync(legacy)) return { path: legacy, mode: 'legacy' };
  return { path: inSecrets, mode: 'secrets' };
}

/** Tóm tắt nguồn của mọi bí mật — an toàn để trả về diagnostics (không có giá trị). */
function describeSecrets() {
  const secretsData = readJsonObject(secretsFilePath());
  const legacyConfig = readJsonObject(LEGACY_CONFIG_PATH);
  const { secrets, files } = loadManifest();
  return {
    secrets_dir: secretsDir(),
    secrets_file_exists: fs.existsSync(secretsFilePath()),
    secrets: secrets.map(spec => {
      const { value, source } = resolveSecret(spec.key, { secretsData, legacyConfig });
      return { key: spec.key, configured: Boolean(value), source };
    }),
    files: files.map(spec => {
      const info = resolveSecretFile(spec.name);
      return { name: spec.name, mode: info.mode, exists: fs.existsSync(info.path) };
    }),
  };
}

module.exports = {
  ROOT_DIR,
  MANIFEST_PATH,
  LEGACY_CONFIG_PATH,
  SOURCE,
  loadManifest,
  secretsDir,
  secretsFilePath,
  readJsonObject,
  getPath,
  setPath,
  resolveSecret,
  getSecret,
  resolveSecretFile,
  describeSecrets,
};
